src/dft/backends/cufft/commit.cpp

```diff
@@ -323,6 +350,77 @@ class cufft_commit final : public dft::detail::commit_impl<prec, dom> {
         return offsets;
     }
 
+    virtual void set_workspace(scalar_type* usm_workspace) override {
+        this->external_workspace_helper_.set_workspace_throw(*this, usm_workspace);
+        if (plans[0]) {
+            cufftSetWorkArea(*plans[0], usm_workspace);
+        }
+        if (plans[1]) {
+            cufftSetWorkArea(
+                *plans[1], usm_workspace + get_plan0_workspace_size_bytes() / sizeof(scalar_type));
+        }
+    }
+
+    void set_buffer_workspace(cufftHandle plan, sycl::buffer<scalar_type>& buffer_workspace,
+                              std::size_t range, std::size_t offset) {
+        this->get_queue()
+            .submit([&](sycl::handler& cgh) {
+                auto workspace_acc =
+                    buffer_workspace.template get_access<sycl::access::mode::read_write>(cgh, range,
+                                                                                         offset);
+                cgh.host_task([=](sycl::interop_handle ih) {
+                    auto stream = ih.get_native_queue<sycl::backend::ext_oneapi_cuda>();
+                    auto result = cufftSetStream(plan, stream);
+                    if (result != CUFFT_SUCCESS) {
+                        throw oneapi::mkl::exception(
+                            "dft/backends/cufft", "set_workspace",
+                            "cufftSetStream returned " + std::to_string(result));
+                    }
+                    auto workspace_native = reinterpret_cast<scalar_type*>(
+                        ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(workspace_acc));
+                    cufftSetWorkArea(plan, workspace_native);
```

Comment:
I do not think this is OK. SYCL runtime only expects the memory to be used within the host_task call. This could lead to multiple subtle bugs. So accessor should be constructed within the same host_task call that launches the cufft kernel.

Reply:
You probably correct from the standpoint of the spec (although the interop spec for CUDA or AMD are not yet published). My understanding is that this should work with Codeplay's oneAPI for Nvidia GPUs and oneAPI for AMD GPUs.

Reply:
I was able to do this more tidily than expected in [fe3273f](https://github.com/oneapi-src/oneMKL/pull/411/commits/fe3273f53bb6b3a0449e94b65bba4d77877a62ee).